packages/rxjs/src/internal/observable/dom/WebSocketSubject.ts

```diff
@@ -151,47 +148,53 @@ const WEBSOCKETSUBJECT_INVALID_ERROR_OBJECT =
 
 export type WebSocketMessage = string | ArrayBuffer | Blob | ArrayBufferView;
 
-export class WebSocketSubject<T> extends AnonymousSubject<T> {
-  private _config!: WebSocketSubjectConfig<T>;
-
-  /** @internal */
-  _output!: Subject<T>;
+export class WebSocketSubject<In, Out = In> extends Observable<Out> {
+  private _config: WebSocketSubjectConfig<In, Out> &
+    Required<Pick<WebSocketSubjectConfig<In, Out>, 'WebSocketCtor' | 'serializer' | 'deserializer'>>;
 
   private _socket: WebSocket | null = null;
 
-  constructor(urlConfigOrSource: string | WebSocketSubjectConfig<T> | Observable<T>, destination?: Observer<T>) {
+  private _inputBuffer: In[] = [];
+
+  private _hasError = false;
+
+  private _error: any;
+
+  private _isComplete = false;
+
+  private _subscriberCounter = 0;
+
+  private _subscribers = new Map<number, Subscriber<Out>>();
+
+  get observed() {
+    return this._subscribers.size > 0;
+  }
+
+  constructor(urlConfigOrSource: string | WebSocketSubjectConfig<In, Out>) {
     super();
-    if (urlConfigOrSource instanceof Observable) {
-      this.destination = destination;
-      this._source = urlConfigOrSource as Observable<T>;
-    } else {
-      const config = (this._config = { ...DEFAULT_WEBSOCKET_CONFIG });
-      this._output = new Subject<T>();
-      if (typeof urlConfigOrSource === 'string') {
-        config.url = urlConfigOrSource;
-      } else {
-        for (const key in urlConfigOrSource) {
-          if (urlConfigOrSource.hasOwnProperty(key)) {
-            (config as any)[key] = (urlConfigOrSource as any)[key];
-          }
-        }
-      }
+    const userConfig = typeof urlConfigOrSource === 'string' ? { url: urlConfigOrSource } : urlConfigOrSource;
+    this._config = {
+      ...DEFAULT_WEBSOCKET_CONFIG,
+      // Setting this here because a previous version of this allowed
+      // WebSocket to be polyfilled later than DEFAULT_WEBSOCKET_CONFIG
+      // was defined.
+      WebSocketCtor: WebSocket,
```

Comment:
Will this throw an exception if the `WebSocket` is not available globally?

`WebSocketCtor` becomes almost useless.